docs/reference/c++_support.rst

```diff
@@ -0,0 +1,165 @@
+
+The C++11 standard introduced a myriad of new features to the language. These features
+are supported in HIP device code, with some notable omissions. The biggest of which is
+the lack of concurrency support on the device. This is because the HIP device concurrency
+model is fundamentally different compared to the C++ model, which is used on the host
+side. E.G: it doesn't make sense to start a new thread on the device.
+
+There are some restrictions and clarifications for certain features. For example, some
+new features were like functions using ``initializer lists``, ``std::move`` and
+``std::forward`` or functions with ``constexpr`` qualifier are implicitly considered to
+have ``__host__`` and ``__device__`` execution space specifier. Also, ``constexpr``
+variables that are static member variables or namespace scoped can be used from both host
+and device, but only for value access. Dereferencing a static ``constexpr`` on a different
+execution space than specified will cause an error.
+
+Lambdas work, but there are some extensions and restrictions on their usage. To learn
+more about them, read the `extended lambdas`_ sections.
+
+C++14 support
+-------------------------------------------------------------------------------
+
+The C++14 language features are supported, except for the updates to the memory model.
+
+C++17 support
+-------------------------------------------------------------------------------
+
+All C++17 language features are supported.
+
+C++20 support
+-------------------------------------------------------------------------------
+
+All C++20 language features are supported, but extensions and restrictions apply. C++20
+introduced features that change programs foundationally with coroutines and modules, but
+unfortunately these aren't supported by HIP. On the other hand, ``consteval`` functions
+can be called from host and device, even if it is specified for host use only. 
+
+The three-way comparison operator (also known as spaceship operator ``<=>``) works in
+both host and device code.
+
+.. _language_restrictions:
+Extensions and Restrictions
+===============================================================================
+
+Besides the above mentioned deviations from the standard, there are more general
+extensions and restrictions to consider. 
+
+``__global__`` functions
+-------------------------------------------------------------------------------
+
+Functions that work as an entry point for device execution are called kernels and are
+specified with the ``__global__`` qualifier. Calling these functions happen with the
+triple chevron operator: ``<<< >>>``. Kernel functions return type must be ``void``, they
+can't be ``constexpr``, can't have a parameter of type ``std::initializer_list`` or
+``va_list``, can't have a parameter of rvalue reference type. Kernels can have variadic
+template parameters, but only one pack, and it must be the last in the template parameter
+list. 
```

Comment:
Functions that serve as an entry point for device execution are called kernels and are specified with the ``__global__`` qualifier. To call a kernel function, use the triple chevron operator: ``<<< >>>``. Kernel functions must have a ``void`` return type. These functions can’t:

•	have a ``constexpr`` specifier
•	have a parameter of type ``std::initializer_list`` or ``va_list``
•	use an rvalue reference as a parameter. 

Kernels can have variadic template parameters, but they can only have one parameter pack, which must be the last item in the template parameter list.